vector/src/main/java/im/vector/app/core/utils/ToolbarConfig.kt

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (c) 2022 New Vector Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package im.vector.app.core.utils
+
+import androidx.annotation.DrawableRes
+import androidx.annotation.StringRes
+import androidx.appcompat.app.AppCompatActivity
+import androidx.appcompat.content.res.AppCompatResources
+import com.google.android.material.appbar.MaterialToolbar
+
+class ToolbarConfig(val activity: AppCompatActivity?, val toolbar: MaterialToolbar) {
+    private var titleRes: Int? = null
+    private var title: String? = null
+    private var subtitleRes: Int? = null
+    private var subtitle: String? = null
+    private var customBackResId: Int? = null
+    private var isBackAllowed: Boolean = true
+    private var useCloseButton: Boolean = false
+
+    fun withTitle(@StringRes titleRes: Int) = apply {this.titleRes = titleRes }
+    fun withTitle(title: String) = apply {this.title = title}
+    fun withSubtitle(@StringRes subtitleRes: Int) = apply {this.subtitleRes = subtitleRes }
+    fun withSubtitle(subtitle: String) = apply {this.subtitle = subtitle}
+
+    fun allowBack(isBackAllowed: Boolean) = apply { this.isBackAllowed = isBackAllowed }
+    fun setBackAsClose(useCloseButton: Boolean) = apply { this.useCloseButton = useCloseButton }
+    fun withCustomBackIcon(@DrawableRes customBackResId: Int) = apply { this.customBackResId = customBackResId }
+
+    fun configure(){
```

Comment:
we could split this responsibilities of this class by moving the config to a data class with a kotlin dsl style builder and then move the toolbar configuration to the activity (or extracted to a delegate)

dsl builders have the benefit of reducing boilerplate 

```kotlin
data class ToolbarConfig(
  var titleRes: Int? = null,
  var allowBack: Boolean = true,
)

// usage
activity.configureToolbar {
    titleRes = R.string.contacts_book_title
    allowBack = true
    customBackIcon = R.drawable.ic_x_18dp
}

fun VectorBaseActivity<*>.configureToolbar(builder: ToolbarConfig.() -> Unit) {
    val config = ToolbarConfig().apply { builder(this) }

    // apply the config to the toolbar
    setSupportActionBar(toolbar)
    toolbar.title = config.titleRes?.let { getString(it) }
    ...
}

```

what do you think?

Reply:
Check new implementation. It's not exactly same, but also reduce amount of boilerplate